### DISCUSSION TITLE
Allow for inlay hints to be automatically disabled in insert mode

### DISCUSSION THREAD

**Discussion by David-Aguilo** [neovim/neovim#29078] [Q&A]
### Problem

Currently, inlay hints remain enabled when entering insert mode, as far as I am aware this behaviour cannot be configured. This causes the cursor to jump around as the type is being inferred when declaring variables, which is in my opinion annoying.

### Expected behavior

When configured, inlay hints are automatically disabled when entering insert mode and re-enabled when insert mode is exited so that when typing code the cursor does not jump around.

**Comment by clason** [comment]
Just use an `InsertEnter` (and `InsertLeave`) autocommand that toggles them? Inlay hints are not enabled by default, so there is nothing to configure in this sense.

**Reply by David-Aguilo** [reply]
Yes, I know that that is an option, I just thought it may be worth making a config option for it, so that it is simpler to setup. Only needing to set something to true instead of writing autocmd.

**Comment by yochem** [comment]
Just for the lazy:

```lua
vim.api.nvim_create_autocmd({ 'InsertEnter', 'InsertLeave' }, {
  callback = function()
    vim.lsp.inlay_hint.enable(not vim.lsp.inlay_hint.is_enabled())
  end
})
```

Or better, to make sure it _disables_ in insert and _enables_ in other modes:

```lua
vim.api.nvim_create_autocmd('InsertEnter', {
  callback = function()
    vim.lsp.inlay_hint.enable(false)
  end
})

vim.api.nvim_create_autocmd('InsertLeave', {
  callback = function()
    vim.lsp.inlay_hint.enable(true)
  end
})
```

**Reply by matta** [reply]
I just put this in a new `plugin/disable-inlay-hint-in-insert-mode.lua` file. Seems to work (I'm using NVIM v10.0.0). If I have inlay hints off, it won't turn them back on when I leave insert mode. If they are on, they will toggle off while in insert mode.

```lua
local was_enabled = true
vim.api.nvim_create_autocmd("InsertEnter", {
  callback = function()
    was_enabled = vim.lsp.inlay_hint.is_enabled({ bufnr = 0 })
    if was_enabled then
      vim.lsp.inlay_hint.enable(false, { bufnr = 0 })
    end
  end,
})

vim.api.nvim_create_autocmd("InsertLeave", {
  callback = function()
    vim.lsp.inlay_hint.enable(was_enabled, { bufnr = 0 })
  end,
})
```

**Reply by yochem** [reply]
I don't think it's possible to update local lua variables like that. If you really want to use a globally accessible variable, use `vim.g.was_enabled = true` (declare without `local`). Otherwise, use the solution I posted.

**Reply by matta** [reply]
> I don't think it's possible to update local lua variables like that.

It seemed to work. A Lua file is also a "chunk" which is treated by Lua as the body of an anonymous function. This means it can have local variables that are shared among the functions that follow in the file. See https://www.lua.org/manual/5.1/manual.html#2.4.1 and https://www.lua.org/manual/5.1/manual.html#2.3

> If you really want to use a globally accessible variable, use vim.g.was_enabled = true (declare without local). Otherwise, use the solution I posted.

I think using a `local` file scoped variable like I did, and `vim.g.was_enabled`, are both wrong. The insert state is a buffer local property, so `vim.b.<varname>` is better.

Instead of that, I decided to use a "once" autocmd for "InsertLeave", so it fires only if the "InsertEnter" actually disabled inlay hinting.

```lua
vim.api.nvim_create_autocmd("InsertEnter", {
  desc = "Disable lsp.inlay_hint when in insert mode",
  callback = function(args)
    local filter = { bufnr = args.buf }
    local inlay_hint = vim.lsp.inlay_hint
    if inlay_hint.is_enabled(filter) then
      inlay_hint.enable(false, filter)
      vim.api.nvim_create_autocmd("InsertLeave", {
        once = true,
        desc = "Re-enable lsp.inlay_hint when leaving insert mode",
        callback = function()
          inlay_hint.enable(true, filter)
        end,
      })
    end
  end,
})

```

**Comment by xzbdmw** [comment]
It is not possible to keep them visible in insert mode but also "freeze" them so it does not make cursor jump around when typing. Have to modify source code for that.
